File: water_reminder.go
# water_reminder.go
/**
 * 💧 Water Reminder – Smart Hydration Tracker with Alerts (Go Edition)
 * Features: add intake, daily goal, configurable reminders, stats, persistence
 */

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// ─── Data Model ─────────────────────────────────────────────────────────────

type Entry struct {
	Date      string `json:"date"`
	Amount    int    `json:"amount"`
	Timestamp string `json:"timestamp"`
}

type Data struct {
	Goal             int     `json:"goal"`
	ReminderInterval int     `json:"reminder_interval"`
	LastReminder     string  `json:"last_reminder"`
	Entries          []Entry `json:"entries"`
}

// ─── Colors ──────────────────────────────────────────────────────────────────

const (
	reset  = "\x1b[0m"
	bright = "\x1b[1m"
	dim    = "\x1b[2m"
	red    = "\x1b[31m"
	green  = "\x1b[32m"
	yellow = "\x1b[33m"
	blue   = "\x1b[34m"
	magenta = "\x1b[35m"
	cyan   = "\x1b[36m"
)

func c(str, color string) string {
	return color + str + reset
}

// ─── Config ──────────────────────────────────────────────────────────────────

const (
	defaultGoal     = 2000
	defaultInterval = 30
)

// ─── Data Manager ──────────────────────────────────────────────────────────

type WaterReminder struct {
	goal         int
	interval     int
	lastReminder string
	entries      []Entry
	filePath     string
	reader       *bufio.Reader
}

func NewWaterReminder() *WaterReminder {
	home, _ := os.UserHomeDir()
	dir := filepath.Join(home, ".water_reminder")
	os.MkdirAll(dir, 0755)
	filePath := filepath.Join(dir, "data.json")
	w := &WaterReminder{filePath: filePath, reader: bufio.NewReader(os.Stdin)}
	w.load()
	w.checkReminder()
	// Periodic check every minute
	go func() {
		ticker := time.NewTicker(1 * time.Minute)
		for range ticker.C {
			w.checkReminder()
		}
	}()
	return w
}

func (w *WaterReminder) load() {
	if _, err := os.Stat(w.filePath); os.IsNotExist(err) {
		w.goal = defaultGoal
		w.interval = defaultInterval
		w.lastReminder = ""
		w.entries = []Entry{}
		return
	}
	raw, err := os.ReadFile(w.filePath)
	if err != nil {
		w.goal = defaultGoal
		w.interval = defaultInterval
		w.lastReminder = ""
		w.entries = []Entry{}
		return
	}
	var data Data
	if err := json.Unmarshal(raw, &data); err != nil {
		w.goal = defaultGoal
		w.interval = defaultInterval
		w.lastReminder = ""
		w.entries = []Entry{}
		return
	}
	w.goal = data.Goal
	if w.goal <= 0 {
		w.goal = defaultGoal
	}
	w.interval = data.ReminderInterval
	if w.interval <= 0 {
		w.interval = defaultInterval
	}
	w.lastReminder = data.LastReminder
	w.entries = data.Entries
	if w.entries == nil {
		w.entries = []Entry{}
	}
}

func (w *WaterReminder) save() {
	data := Data{
		Goal:             w.goal,
		ReminderInterval: w.interval,
		LastReminder:     w.lastReminder,
		Entries:          w.entries,
	}
	raw, _ := json.MarshalIndent(data, "", "  ")
	os.WriteFile(w.filePath, raw, 0644)
}

func (w *WaterReminder) today() string {
	return time.Now().Format("2006-01-02")
}

func (w *WaterReminder) getTodayEntries() []Entry {
	today := w.today()
	var res []Entry
	for _, e := range w.entries {
		if e.Date == today {
			res = append(res, e)
		}
	}
	return res
}

func (w *WaterReminder) getTodayTotal() int {
	total := 0
	for _, e := range w.getTodayEntries() {
		total += e.Amount
	}
	return total
}

func (w *WaterReminder) progressBar(current, goal, width int) string {
	if goal <= 0 {
		return "⚠️  Goal not set"
	}
	ratio := float64(current) / float64(goal)
	if ratio > 1.0 {
		ratio = 1.0
	}
	filled := int(ratio * float64(width))
	bar := strings.Repeat("█", filled) + strings.Repeat("░", width-filled)
	return fmt.Sprintf("[%s] %.1f%%", bar, ratio*100)
}

func (w *WaterReminder) checkReminder() {
	if w.lastReminder == "" {
		w.lastReminder = time.Now().Format(time.RFC3339)
		w.save()
		return
	}
	last, err := time.Parse(time.RFC3339, w.lastReminder)
	if err != nil {
		w.lastReminder = time.Now().Format(time.RFC3339)
		w.save()
		return
	}
	elapsed := time.Since(last).Minutes()
	if elapsed >= float64(w.interval) {
		fmt.Printf("%s\n", c("\n⏰ Time to drink water! (%d min since last reminder)", cyan), w.interval)
		w.lastReminder = time.Now().Format(time.RFC3339)
		w.save()
	}
}

// ─── Core Actions ──────────────────────────────────────────────────────────

func (w *WaterReminder) addEntry(amount int) {
	if amount <= 0 {
		fmt.Println(c("❌ Amount must be positive!", red))
		return
	}
	w.entries = append(w.entries, Entry{
		Date:      w.today(),
		Amount:    amount,
		Timestamp: time.Now().Format(time.RFC3339),
	})
	w.save()
	todayTotal := w.getTodayTotal()
	fmt.Printf(c("✅ Added %dml (Total today: %dml)\n", green), amount, todayTotal)
	if todayTotal >= w.goal {
		fmt.Println(c("🎉 Goal achieved! Stay hydrated! 💪", cyan))
	}
}

func (w *WaterReminder) showToday() {
	todayTotal := w.getTodayTotal()
	entries := w.getTodayEntries()
	fmt.Println("\n" + c(strings.Repeat("═", 50), dim))
	fmt.Println(c("💧 TODAY'S HYDRATION", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), dim))
	fmt.Printf("  Goal:      %s\n", c(strconv.Itoa(w.goal)+"ml", cyan))
	fmt.Printf("  Consumed:  %s\n", c(strconv.Itoa(todayTotal)+"ml", green))
	remaining := w.goal - todayTotal
	if remaining < 0 {
		remaining = 0
	}
	fmt.Printf("  Remaining: %s\n", c(strconv.Itoa(remaining)+"ml", yellow))
	fmt.Printf("  Progress:  %s\n", w.progressBar(todayTotal, w.goal, 20))
	fmt.Println(c(strings.Repeat("═", 50), dim))
	if len(entries) > 0 {
		fmt.Println("  Entries:")
		for i, e := range entries {
			ts := "—"
			if len(e.Timestamp) >= 16 {
				ts = e.Timestamp[11:16]
			}
			fmt.Printf("    %d. %s → %s\n", i+1, ts, c(strconv.Itoa(e.Amount)+"ml", green))
		}
	} else {
		fmt.Println(c("  No entries yet today. Drink up! 💧", dim))
	}
}

func (w *WaterReminder) showStats() {
	if len(w.entries) == 0 {
		fmt.Println(c("📭 No data yet. Start tracking!", yellow))
		return
	}
	total := 0
	for _, e := range w.entries {
		total += e.Amount
	}
	count := len(w.entries)
	avg := float64(total) / float64(count)
	days := make(map[string]bool)
	for _, e := range w.entries {
		days[e.Date] = true
	}
	fmt.Println("\n📊 STATISTICS")
	fmt.Println(c(strings.Repeat("─", 30), dim))
	fmt.Printf("  Total Consumed: %dml\n", total)
	fmt.Printf("  Total Entries:  %d\n", count)
	fmt.Printf("  Days Tracked:   %d\n", len(days))
	fmt.Printf("  Average per Entry: %.1fml\n", avg)
	fmt.Printf("  Daily Goal:     %dml\n", w.goal)
	fmt.Printf("  Reminder Interval: %d min\n", w.interval)
}

func (w *WaterReminder) setGoal(goal int) {
	if goal <= 0 {
		fmt.Println(c("❌ Goal must be positive!", red))
		return
	}
	w.goal = goal
	w.save()
	fmt.Printf(c("✅ Daily goal set to %dml\n", green), goal)
}

func (w *WaterReminder) setInterval(minutes int) {
	if minutes <= 0 {
		fmt.Println(c("❌ Interval must be positive!", red))
		return
	}
	w.interval = minutes
	w.save()
	fmt.Printf(c("✅ Reminder interval set to %d minutes\n", green), minutes)
}

func (w *WaterReminder) remindNow() {
	fmt.Println(c("\n💧 Time to drink some water! Stay hydrated!", cyan))
	w.lastReminder = time.Now().Format(time.RFC3339)
	w.save()
}

func (w *WaterReminder) clearData() {
	fmt.Print("⚠️  Delete ALL data? (yes/no): ")
	ans, _ := w.reader.ReadString('\n')
	ans = strings.TrimSpace(strings.ToLower(ans))
	if ans != "yes" {
		return
	}
	w.entries = []Entry{}
	w.goal = defaultGoal
	w.interval = defaultInterval
	w.lastReminder = ""
	w.save()
	fmt.Println(c("🗑️  All data cleared.", yellow))
}

// ─── Menu ──────────────────────────────────────────────────────────────────

func (w *WaterReminder) ask(prompt string) string {
	fmt.Print(prompt)
	line, _ := w.reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func (w *WaterReminder) askInt(prompt string) int {
	for {
		ans := w.ask(prompt)
		if val, err := strconv.Atoi(ans); err == nil {
			return val
		}
		fmt.Println(c("❌ Please enter a number.", red))
	}
}

func (w *WaterReminder) showMenu() {
	todayTotal := w.getTodayTotal()
	progress := w.progressBar(todayTotal, w.goal, 20)
	fmt.Println("\n" + c(strings.Repeat("═", 50), cyan))
	fmt.Println(c("💧 WATER REMINDER", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Printf("  Today: %dml / %dml  %s\n", todayTotal, w.goal, progress)
	fmt.Printf("  Reminder: every %d min\n", w.interval)
	fmt.Println(c(strings.Repeat("─", 50), dim))
	fmt.Println("  1. 💧 Add water intake")
	fmt.Println("  2. 📊 Today's progress")
	fmt.Println("  3. 📈 Statistics")
	fmt.Printf("  4. 🎯 Set daily goal (current: %dml)\n", w.goal)
	fmt.Printf("  5. ⏰ Set reminder interval (current: %d min)\n", w.interval)
	fmt.Println("  6. 🔔 Check reminder now")
	fmt.Println("  7. 🗑️  Clear all data")
	fmt.Println("  0. 🚪 Exit")
	fmt.Println(c(strings.Repeat("═", 50), cyan))
}

func (w *WaterReminder) run() {
	fmt.Print("\033[H\033[2J")
	fmt.Println(c("\n💧 Water Reminder – Stay Hydrated!", bright+cyan))
	fmt.Println(c("Never forget to drink water again!", dim))

	for {
		w.showMenu()
		choice := w.ask("Your choice: ")
		switch choice {
		case "1":
			amount := w.askInt("Amount in ml: ")
			w.addEntry(amount)
		case "2":
			w.showToday()
		case "3":
			w.showStats()
		case "4":
			goal := w.askInt("New daily goal (ml): ")
			w.setGoal(goal)
		case "5":
			interval := w.askInt("Interval (minutes): ")
			w.setInterval(interval)
		case "6":
			w.remindNow()
		case "7":
			w.clearData()
		case "0":
			fmt.Println(c("👋 Stay hydrated! Goodbye!", cyan))
			return
		default:
			fmt.Println(c("❌ Invalid choice.", red))
		}
		if choice != "0" {
			fmt.Print("\nPress Enter to continue...")
			w.reader.ReadString('\n')
		}
	}
}

func main() {
	app := NewWaterReminder()
	app.run()
}

File: water_reminder.rs
# water_reminder.rs
/**
 * 💧 Water Reminder – Smart Hydration Tracker with Alerts (Rust Edition)
 * Features: add intake, daily goal, configurable reminders, stats, persistence
 * Dependencies: serde, serde_json, chrono, colored
 */

use chrono::{DateTime, Local, Duration};
use colored::*;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::PathBuf;
use std::thread;
use std::time;

// ─── Data Model ─────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
struct Entry {
    date: String,
    amount: u32,
    timestamp: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct Data {
    goal: u32,
    reminder_interval: u32,
    last_reminder: String,
    entries: Vec<Entry>,
}

// ─── Colors ──────────────────────────────────────────────────────────────────

fn c(text: &str, color: &str) -> String {
    match color {
        "green" => text.green().to_string(),
        "red" => text.red().to_string(),
        "yellow" => text.yellow().to_string(),
        "cyan" => text.cyan().to_string(),
        "bright" => text.bright().to_string(),
        "dim" => text.dimmed().to_string(),
        _ => text.to_string(),
    }
}

// ─── Config ──────────────────────────────────────────────────────────────────

const DEFAULT_GOAL: u32 = 2000;
const DEFAULT_INTERVAL: u32 = 30;

// ─── Data Manager ──────────────────────────────────────────────────────────

struct WaterReminder {
    goal: u32,
    interval: u32,
    last_reminder: String,
    entries: Vec<Entry>,
    file_path: PathBuf,
}

impl WaterReminder {
    fn new() -> Self {
        let home = std::env::var("HOME").or_else(|_| std::env::var("USERPROFILE")).unwrap_or_else(|_| ".".to_string());
        let dir = PathBuf::from(home).join(".water_reminder");
        fs::create_dir_all(&dir).unwrap();
        let file_path = dir.join("data.json");
        let mut w = WaterReminder {
            goal: DEFAULT_GOAL,
            interval: DEFAULT_INTERVAL,
            last_reminder: "".to_string(),
            entries: Vec::new(),
            file_path,
        };
        w.load();
        w.check_reminder();
        // Spawn a thread for periodic reminder check
        let w_clone = w.clone_without_io();
        thread::spawn(move || {
            let interval = time::Duration::from_secs(60);
            loop {
                thread::sleep(interval);
                w_clone.check_reminder();
            }
        });
        w
    }

    // Helper to clone without the file_path (we can't share easily)
    fn clone_without_io(&self) -> Self {
        WaterReminder {
            goal: self.goal,
            interval: self.interval,
            last_reminder: self.last_reminder.clone(),
            entries: self.entries.clone(),
            file_path: self.file_path.clone(),
        }
    }

    fn load(&mut self) {
        if let Ok(raw) = fs::read_to_string(&self.file_path) {
            if let Ok(data) = serde_json::from_str::<Data>(&raw) {
                self.goal = if data.goal > 0 { data.goal } else { DEFAULT_GOAL };
                self.interval = if data.reminder_interval > 0 { data.reminder_interval } else { DEFAULT_INTERVAL };
                self.last_reminder = data.last_reminder;
                self.entries = data.entries;
                return;
            }
        }
        self.goal = DEFAULT_GOAL;
        self.interval = DEFAULT_INTERVAL;
        self.last_reminder = "".to_string();
        self.entries = Vec::new();
    }

    fn save(&self) {
        let data = Data {
            goal: self.goal,
            reminder_interval: self.interval,
            last_reminder: self.last_reminder.clone(),
            entries: self.entries.clone(),
        };
        let raw = serde_json::to_string_pretty(&data).unwrap();
        let _ = fs::write(&self.file_path, raw);
    }

    fn today(&self) -> String {
        Local::now().format("%Y-%m-%d").to_string()
    }

    fn get_today_entries(&self) -> Vec<Entry> {
        let today = self.today();
        self.entries.iter().filter(|e| e.date == today).cloned().collect()
    }

    fn get_today_total(&self) -> u32 {
        self.get_today_entries().iter().map(|e| e.amount).sum()
    }

    fn progress_bar(&self, current: u32, goal: u32, width: usize) -> String {
        if goal == 0 {
            return "⚠️  Goal not set".to_string();
        }
        let ratio = (current as f64 / goal as f64).min(1.0);
        let filled = (ratio * width as f64) as usize;
        let bar = "█".repeat(filled) + &"░".repeat(width - filled);
        format!("[{}] {:.1}%", bar, ratio * 100.0)
    }

    fn check_reminder(&self) {
        if self.last_reminder.is_empty() {
            // first run, set last reminder to now
            let mut w = self.clone_without_io();
            w.last_reminder = Local::now().to_rfc3339();
            w.save();
            return;
        }
        if let Ok(last) = DateTime::parse_from_rfc3339(&self.last_reminder) {
            let now = Local::now();
            let elapsed = (now - last).num_minutes();
            if elapsed >= self.interval as i64 {
                println!("{}", c(&format!("\n⏰ Time to drink water! ({} min since last reminder)", self.interval), "cyan"));
                let mut w = self.clone_without_io();
                w.last_reminder = now.to_rfc3339();
                w.save();
            }
        }
    }

    // ─── Core Actions ──────────────────────────────────────────────────────

    fn add_entry(&mut self, amount: u32) {
        if amount == 0 {
            println!("{}", c("❌ Amount must be positive!", "red"));
            return;
        }
        self.entries.push(Entry {
            date: self.today(),
            amount,
            timestamp: Local::now().to_rfc3339(),
        });
        self.save();
        let today_total = self.get_today_total();
        println!("{}", c(&format!("✅ Added {}ml (Total today: {}ml)", amount, today_total), "green"));
        if today_total >= self.goal {
            println!("{}", c("🎉 Goal achieved! Stay hydrated! 💪", "cyan"));
        }
    }

    fn show_today(&self) {
        let today_total = self.get_today_total();
        let entries = self.get_today_entries();
        println!("\n{}", "═".repeat(50).dimmed());
        println!("{}", c("💧 TODAY'S HYDRATION", "bright") + &c("", "cyan"));
        println!("{}", "═".repeat(50).dimmed());
        println!("  Goal:      {}", c(&format!("{}ml", self.goal), "cyan"));
        println!("  Consumed:  {}", c(&format!("{}ml", today_total), "green"));
        let remaining = if self.goal > today_total { self.goal - today_total } else { 0 };
        println!("  Remaining: {}", c(&format!("{}ml", remaining), "yellow"));
        println!("  Progress:  {}", self.progress_bar(today_total, self.goal, 20));
        println!("{}", "═".repeat(50).dimmed());
        if entries.is_empty() {
            println!("{}", c("  No entries yet today. Drink up! 💧", "dim"));
        } else {
            println!("  Entries:");
            for (i, e) in entries.iter().enumerate() {
                let ts = if e.timestamp.len() >= 16 { &e.timestamp[11..16] } else { "—" };
                println!("    {}. {} → {}", i+1, ts, c(&format!("{}ml", e.amount), "green"));
            }
        }
    }

    fn show_stats(&self) {
        if self.entries.is_empty() {
            println!("{}", c("📭 No data yet. Start tracking!", "yellow"));
            return;
        }
        let total: u32 = self.entries.iter().map(|e| e.amount).sum();
        let count = self.entries.len();
        let avg = total as f64 / count as f64;
        let days: HashSet<_> = self.entries.iter().map(|e| &e.date).collect();
        println!("\n📊 STATISTICS");
        println!("{}", "─".repeat(30).dimmed());
        println!("  Total Consumed: {}ml", total);
        println!("  Total Entries:  {}", count);
        println!("  Days Tracked:   {}", days.len());
        println!("  Average per Entry: {:.1}ml", avg);
        println!("  Daily Goal:     {}ml", self.goal);
        println!("  Reminder Interval: {} min", self.interval);
    }

    fn set_goal(&mut self, goal: u32) {
        if goal == 0 {
            println!("{}", c("❌ Goal must be positive!", "red"));
            return;
        }
        self.goal = goal;
        self.save();
        println!("{}", c(&format!("✅ Daily goal set to {}ml", goal), "green"));
    }

    fn set_interval(&mut self, minutes: u32) {
        if minutes == 0 {
            println!("{}", c("❌ Interval must be positive!", "red"));
            return;
        }
        self.interval = minutes;
        self.save();
        println!("{}", c(&format!("✅ Reminder interval set to {} minutes", minutes), "green"));
    }

    fn remind_now(&mut self) {
        println!("{}", c("\n💧 Time to drink some water! Stay hydrated!", "cyan"));
        self.last_reminder = Local::now().to_rfc3339();
        self.save();
    }

    fn clear_data(&mut self) {
        print!("⚠️  Delete ALL data? (yes/no): ");
        io::stdout().flush().unwrap();
        let mut ans = String::new();
        io::stdin().read_line(&mut ans).unwrap();
        if ans.trim().to_lowercase() != "yes" { return; }
        self.entries = Vec::new();
        self.goal = DEFAULT_GOAL;
        self.interval = DEFAULT_INTERVAL;
        self.last_reminder = "".to_string();
        self.save();
        println!("{}", c("🗑️  All data cleared.", "yellow"));
    }

    // ─── Menu ──────────────────────────────────────────────────────────────

    fn ask(&self, prompt: &str) -> String {
        print!("{}", prompt);
        io::stdout().flush().unwrap();
        let mut line = String::new();
        io::stdin().read_line(&mut line).unwrap();
        line.trim().to_string()
    }

    fn ask_u32(&self, prompt: &str) -> u32 {
        loop {
            let ans = self.ask(prompt);
            if let Ok(val) = ans.parse::<u32>() {
                return val;
            }
            println!("{}", c("❌ Please enter a number.", "red"));
        }
    }

    fn show_menu(&self) {
        let today_total = self.get_today_total();
        let progress = self.progress_bar(today_total, self.goal, 20);
        println!("\n{}", "═".repeat(50).cyan());
        println!("{}", c("💧 WATER REMINDER", "bright") + &c("", "cyan"));
        println!("{}", "═".repeat(50).cyan());
        println!("  Today: {}ml / {}ml  {}", today_total, self.goal, progress);
        println!("  Reminder: every {} min", self.interval);
        println!("{}", "─".repeat(50).dimmed());
        println!("  1. 💧 Add water intake");
        println!("  2. 📊 Today's progress");
        println!("  3. 📈 Statistics");
        println!("  4. 🎯 Set daily goal (current: {}ml)", self.goal);
        println!("  5. ⏰ Set reminder interval (current: {} min)", self.interval);
        println!("  6. 🔔 Check reminder now");
        println!("  7. 🗑️  Clear all data");
        println!("  0. 🚪 Exit");
        println!("{}", "═".repeat(50).cyan());
    }

    fn run(&mut self) {
        println!("{}", "\n💧 Water Reminder – Stay Hydrated!".bright().cyan());
        println!("{}", "Never forget to drink water again!".dimmed());

        loop {
            self.show_menu();
            let choice = self.ask("Your choice: ");
            match choice.as_str() {
                "1" => {
                    let amount = self.ask_u32("Amount in ml: ");
                    self.add_entry(amount);
                }
                "2" => self.show_today(),
                "3" => self.show_stats(),
                "4" => {
                    let goal = self.ask_u32("New daily goal (ml): ");
                    self.set_goal(goal);
                }
                "5" => {
                    let interval = self.ask_u32("Interval (minutes): ");
                    self.set_interval(interval);
                }
                "6" => self.remind_now(),
                "7" => self.clear_data(),
                "0" => {
                    println!("{}", c("👋 Stay hydrated! Goodbye!", "cyan"));
                    return;
                }
                _ => println!("{}", c("❌ Invalid choice.", "red")),
            }
            if choice != "0" {
                print!("\nPress Enter to continue...");
                io::stdout().flush().unwrap();
                let mut _dummy = String::new();
                io::stdin().read_line(&mut _dummy).unwrap();
            }
        }
    }
}

// ─── Main ────────────────────────────────────────────────────────────────────

fn main() {
    let mut app = WaterReminder::new();
    app.run();
}
